File: base/hourai/bot/extensions/verification/approvers.py
from .common import Verifier
from hourai import utils


class NitroApprover(Verifier):
    """A suspicion level verifier that approves users that have Nitro.

    Note: there is currently no way to confirm if a user has Nitro or not
    directly from the API. The only way to confirm it is through peripheral
    effects (i.e. animated avatars), so this verifier has a higher chance of
    false negatives.

    Uses the following attributes only Nitro users have access to:
     - Nitro Boosting: only works on servers shared between the bot and the
       user.
     - Animated Avatars
     - Third-Party Emotes (not implemented)
    """

    async def verify_member(self, ctx):
        if utils.has_nitro(ctx.bot, ctx.member):
            ctx.add_approval_reason('User currently has or has had Nitro. '
                                    'Probably not a user bot.')


class BotApprover(Verifier):
    """A override level verifier that approves other bots."""

    async def verify_member(self, ctx):
        if ctx.member.bot:
            ctx.add_approval_reason(
                'User is an OAuth2 bot that can only be manually added.')


class BotOwnerApprover(Verifier):
    """An override level verifier that approves the owner of the bot or part of
    the team that owns the bot."""

    async def verify_member(self, ctx):
        if (await ctx.bot.is_owner(ctx.member)):
            ctx.add_approval_reason("User owns this bot.")


class DistinguishedUserApprover(Verifier):
    """A malice level approver that approves distinguished users. Approves the
    following:
    - Discord Staff
    - Discord Partners
    - Owners of verified servers*

    Since bot users cannot read profile data, the bot must be on the same
    server for this to work. Any of the designations marked above by a * may
    result in many false negatives.
    """

    FLAG_MATCHES = {
        "staff": 'User is Discord Staff.',
        "partner": 'User is a Discord Partner.',
        "verified_bot_developer": 'User is a verified bot developer.',
    }

    SERVER_SEARCH_MATCHES = {
        "VERIFIED": 'User is owner of verfied server: "{}"'
    }

    async def verify_member(self, ctx):
        self.__verify_via_user_flags(ctx)
        self.__verify_via_server_ownership(ctx)

    def __verify_via_user_flags(self, ctx):
        flags = ctx.member.public_flags
        for attr, reason in self.FLAG_MATCHES.items():
            if getattr(flags, attr):
                ctx.add_approval_reason(reason)

    def __verify_via_server_ownership(self, ctx):
        # FIXME: This will not scale to multiple processes/nodes
        owned_guilds = [guild for guild in ctx.bot.guilds
                        if guild.owner == ctx.member]
        for guild in owned_guilds:
            for check, reason_template in self.SERVER_SEARCH_MATCHES.items():
                if check in guild.features:
                    ctx.add_approval_reason(reason_template.format(guild.name))

File: logger/hourai-verification/src/approvers.rs
use super::{context, verifier::*};
use anyhow::Result;
use async_trait::async_trait;
use hourai::cache::InMemoryCache;
use hourai::models::id::UserId;
use hourai::models::user::{User, UserFlags};
use std::collections::HashSet;

const VERIFIED_FEATURE: &str = "VERIFIED";

struct DistinguishedUserVerifier(InMemoryCache);

#[async_trait]
impl Verifier for DistinguishedUserVerifier {
    async fn verify(&self, ctx: &mut context::VerificationContext) -> Result<()> {
        let flags = ctx.member().user.flags.unwrap_or_else(UserFlags::empty);
        if flags.contains(UserFlags::DISCORD_EMPLOYEE) {
            ctx.add_approval_reason("User is Discord Staff.");
        }
        if flags.contains(UserFlags::DISCORD_PARTNER) {
            ctx.add_approval_reason("User is a Discord Partner.");
        }
        if flags.contains(UserFlags::VERIFIED_BOT_DEVELOPER) {
            ctx.add_approval_reason("User is a verified bot developer.");
        }
        // TODO(james7123): This will not scale to multiple processes
        //let member_id = ctx.member().user.id;
        //for guild_id in self.0.guilds() {
        //if let Some(guild) = self.0.guild(guild_id) {
        //let verified = guild
        //.features
        //.iter()
        //.any(|feat| feat.as_ref() == VERIFIED_FEATURE);
        //if guild.owner_id == member_id && verified {
        //ctx.add_approval_reason("User is the owner of a verified server.");
        //}
        //}
        //}
        Ok(())
    }
}

pub fn user_has_nitro(user: &User) -> bool {
    let flag = user
        .flags
        .map(|f| f.contains(UserFlags::EARLY_SUPPORTER))
        .unwrap_or(false);
    let animated = user
        .avatar
        .as_ref()
        .map(|a| a.starts_with("a_"))
        .unwrap_or(false);
    flag || animated
}

pub(super) fn nitro() -> BoxedVerifier {
    GenericVerifier::new_approver(
        "User currently has or has had Nitro. Probably not a user bot.",
        |ctx| Ok(user_has_nitro(&ctx.member().user)),
    )
}

pub(super) fn bot_owners(owners: impl IntoIterator<Item = UserId>) -> BoxedVerifier {
    let owner_ids: HashSet<UserId> = owners.into_iter().collect();
    GenericVerifier::new_approver("User is an owner of this bot.", move |ctx| {
        Ok(owner_ids.contains(&ctx.member().user.id))
    })
}

pub(super) fn bot() -> BoxedVerifier {
    GenericVerifier::new_approver(
        "User is an OAuth2 bot that can only be manually added by moderators.",
        |ctx| Ok(ctx.member().user.bot),
    )
}

pub(super) fn distinguished_user(cache: InMemoryCache) -> BoxedVerifier {
    Box::new(DistinguishedUserVerifier(cache))
}
